pep-0634.rst

```diff
@@ -0,0 +1,706 @@
+~~~~~~~~~~~~~~
+
+Syntax:
+
+  group_pattern: '(' pattern ')'
+
+(For the syntax of ``pattern``, see Patterns above.)
+
+A parenthesized pattern has no additional syntax.  It allows uses to
+add parentheses around patterns to emphasize the intended grouping.
+
+
+.. _sequence_pattern:
+
+Sequence Patterns
+~~~~~~~~~~~~~~~~~
+
+
+Syntax::
+
+  sequence_pattern:
+    | '[' [values_pattern] ']'
+    | '(' [value_pattern ',' [values_pattern]] ')'
+
+(For ``values_pattern`` and ``value_pattern``, see Patterns above.)
+
+(Note that a single parenthesized pattern without a trailing comma is
+a group pattern, not a sequence pattern.)
+
+A sequence pattern may (directly) contain at most one subpattern of
+the form ``'*' NAME``; all other subpatterns must be walrus patterns,
+OR patterns or closed patterns.
+
+A sequence pattern fails if the subject value is not an instance of
+``collections.abc.Sequence``.  It also fails if the subject value is
+an instance of ``str``, ``bytes`` or ``bytearray``.
+
+If the one of the subpatterns has the form ``'*' NAME``, this is
+called a variable-length sequence pattern.  A variable-length sequence
+pattern fails if the length of the subject sequence is less than the
+number of subpatterns not of that form.
+
+If no such subpattern is present, the sequence pattern is considered
+fixed-length.  A fixed-length sequence pattern fails if the length of
+the subject sequence is not equal to the number of subpatterns.
+
+The length of the subject sequence is obtained using the builtin
+``len()`` function (i.e., via the ``__len__`` protocol).  However, the
+interpreter may cache this value in a similar manner as described for
+constant value patterns.
+
+TODO: Do we have left-to-right semantics here?
+
+A fixed-length sequence pattern matches the subpatterns to
+corresponding items of the subject sequence, from left to right.
+Matching stops (with a failure) as soon as a subpattern fails.  If all
+subpatterns succeed in matching their corresponding item, the sequence
+pattern succeeds.
+
+TODO: Describe variable-length sequence patterns.  (Brandt?)
+
+
+.. _mapping_pattern:
+
+Mapping Patterns
+~~~~~~~~~~~~~~~~
+
+Syntax::
+
+  mapping_pattern: '{' items_pattern? '}'
+  items_pattern: ','.key_value_pattern+ ','?
+  key_value_pattern:
+      | (literal_pattern | constant_pattern) ':' or_pattern
+      | '**' (!"_" NAME)
+
+(Note that ``'**' "_"`` is disallowed by this grammar.)
+
+A mapping pattern fails if the subject value is not an instance of
+``collections.abc.Mapping``.
+
+A mapping pattern succeeds if every key given in the mapping pattern
+matches the corresponding item of the subject mapping.  If a ``'**'
+NAME`` form is present, that name is bound to a ``dict`` containing
+remaining key-value pairs from the subject mapping.
+
+If duplicate keys are detected in the mapping pattern, the pattern is
+considered invalid, and a ``ValueError`` is raised.
+
+TODO: Should we be more prescriptive about using ``in``?
```

Comment:
Hm, maybe not. The current implementation uses `subject.get(key, sentinel)` for all non-dicts (there is a fast path for exact dicts). This lets us do each key-value pair in one logical lookup (`in` would require 2).

If we should prescribe anything, I propose it be `get()`.

Reply:
Okay, I didn’t know that. Please rewrite the section!